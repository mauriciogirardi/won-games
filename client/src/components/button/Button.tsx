'use client'

import {
  AnchorHTMLAttributes,
  ButtonHTMLAttributes,
  ElementType,
  ReactNode
} from 'react'
import * as S from './Button.styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  fullWidth?: boolean
  icon?: ReactNode
  as?: ElementType
} & ButtonTypes

export function Button({
  children,
  size = 'medium',
  fullWidth = false,
  icon,
  ...props
}: ButtonProps) {
  return (
    <S.ButtonContainer
      size={size}
      fullWidth={fullWidth}
      hasIcon={!!icon}
      {...props}
    >
      {!!icon && icon}
      {!!children && <span>{children}</span>}
    </S.ButtonContainer>
  )
}
