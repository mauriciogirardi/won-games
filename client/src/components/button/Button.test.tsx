import { screen } from '@testing-library/react'
import { Button } from './Button'
import { renderWithTheme } from '@/utils/tests/helpers'
import { ShoppingCart } from 'lucide-react'

describe('<Button />', () => {
  it('should render the medium size by default', () => {
    const { container } = renderWithTheme(<Button>Buy now</Button>)
    const button = screen.getByRole('button', { name: /Buy now/i })

    expect(button).toHaveStyle({
      padding: '0.8rem 3.2rem',
      height: '4rem',
      'font-size': '1.4rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the large size', () => {
    renderWithTheme(<Button size="large">Buy now</Button>)
    const button = screen.getByRole('button', { name: /Buy now/i })

    expect(button).toHaveStyle({
      padding: '0.8rem 4.0rem',
      height: '5rem',
      'font-size': '1.6rem'
    })
  })

  it('should render the small size', () => {
    renderWithTheme(<Button size="small">Buy now</Button>)
    const button = screen.getByRole('button', { name: /Buy now/i })

    expect(button).toHaveStyle({
      height: '3rem',
      'font-size': '1.2rem'
    })
  })

  it('should render a fullWidth version', () => {
    renderWithTheme(<Button fullWidth>Buy now</Button>)
    const button = screen.getByRole('button', { name: /Buy now/i })

    expect(button).toHaveStyle({
      width: '100%'
    })
  })

  it('should render an icon version', () => {
    renderWithTheme(
      <Button icon={<ShoppingCart data-testid="icon" />}>Buy now</Button>
    )
    const buttonText = screen.getByText(/Buy now/i)
    const icon = screen.getByTestId('icon')

    expect(buttonText).toBeInTheDocument()
    expect(icon).toBeInTheDocument()
  })

  it('should render button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Buy now
      </Button>
    )

    expect(screen.getByRole('link', { name: /buy now/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })

  it('should render a minimal version', () => {
    renderWithTheme(
      <Button icon={<ShoppingCart data-testid="icon" />} minimal>
        Buy now
      </Button>
    )

    expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
      background: 'none',
      color: 'rgb(178, 11, 111)'
    })

    expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyleRule(
      'background',
      'none',
      {
        modifier: ':hover'
      }
    )
  })
})
