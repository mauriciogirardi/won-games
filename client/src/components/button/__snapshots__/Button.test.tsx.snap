// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size by default 1`] = `
.c0 {
  background: linear-gradient(180deg, #ff5f5f 0%, #f062c0 50%);
  border-radius: 0.4rem;
  border: 0;
  color: #FAFAFA;
  padding: 0.8rem;
  cursor: pointer;
  overflow: hidden;
  text-decoration: none;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
  height: 4rem;
}

.c0:hover {
  background: linear-gradient(180deg, #e35565 0%, #d958a6 50%);
}

<button
  class="c0"
>
  <span>
    Buy now
  </span>
</button>
`;
